Fix handling thread abort in HelperMethodFrame

janvorli opened the pull request:
The thread abort during func eval from a managed debugger on Linux and macOS
was sometimes causing the debuggee to exit with unhandled c++ PAL_SEHException.
The reason is that the thread abort detection that is done in the
HELPER_METHOD_FRAME_BEGIN and ...END macros was done outside of the
INSTALL_MANAGED_EXCEPTION_DISPATCHER / UNINSTALL_MANAGED_EXCEPTION_DISPATCHER
region and so the exception thrown when thread abort request is detected
there was not being caught and translated into a call to DispatchManagedException.
Since the caller frame was a managed function frame, the C++ exception handling
didn't know how to unwind it and so it declared the exception being unhandled.

This fix moves the check for the thread abort out of the HelperMethodFrame::Push/Pop
into a new method and calls that explicitly from the HELPER_METHOD_* macros inside
the INSTALL_MANAGED_EXCEPTION_DISPATCHER / UNINSTALL_MANAGED_EXCEPTION_DISPATCHER
region. That way, the thread abort exception is properly handled.

jkotas commented on src/coreclr/src/vm/frames.cpp:
```diff
@@ -1657,12 +1657,6 @@ void HelperMethodFrame::Push()
     // Push the frame
     Frame::Push(pThread);
-
-    if (!pThread->HasThreadStateOpportunistic(Thread::TS_AbortRequested))
```
Can removing this fast path cause perf regressions?

jkotas replied:
It may be better to just move `INSTALL_MANAGED_EXCEPTION_DISPATCHER` into `HELPER_METHOD_FRAME_BEGIN_EX_BODY` right after `__helperframe.Push();`.

janvorli replied:
@jkotas I've updated it along the lines you've suggested. I have originally wanted to do it this way, but I got worried about moving the actual frame popping inside of the try regions created by the INSTALL_MANAGED_EXCEPTION_DISPATCHER / UNINSTALL_MANAGED_EXCEPTION_DISPATCHER  and INSTALL_UNWIND_AND_CONTINUE_HANDLER / UNINSTALL_UNWIND_AND_CONTINUE_HANDLER.

But after thinking about it, it seems to be fine as it is the last thing that happens in the try region.

janvorli committed:
Fix handling thread abort in HelperMethodFrame

The thread abort during func eval from a managed debugger on Linux and macOS
was sometimes causing the debuggee to exit with unhandled c++ PAL_SEHException.
The reason is that the thread abort detection that is done in the
HELPER_METHOD_FRAME_BEGIN and ...END macros was done outside of the
INSTALL_MANAGED_EXCEPTION_DISPATCHER / UNINSTALL_MANAGED_EXCEPTION_DISPATCHER
region and so the exception thrown when thread abort request is detected
there was not being caught and translated into a call to DispatchManagedException.
Since the caller frame was a managed function frame, the C++ exception handling
didn't know how to unwind it and so it declared the exception being unhandled.

This fix moves the frame pushing and popping inside of the
INSTALL_MANAGED_EXCEPTION_DISPATCHER / UNINSTALL_MANAGED_EXCEPTION_DISPATCHER
region. That way, the thread abort exception is properly handled.

jkotas commented on src/coreclr/src/vm/fcall.h:
```diff
@@ -606,11 +605,13 @@ LPVOID __FCThrowArgument(LPVOID me, enum RuntimeExceptionKind reKind, LPCWSTR ar
         } FORLAZYMACHSTATE_ENDLOOP(alwaysZero);
 
 #define HELPER_METHOD_FRAME_END_EX(gcpoll,allowGC)                          \
+            __helperframe.Pop();                                            \
```
Should this be between `UNINSTALL_UNWIND_AND_CONTINUE_HANDLER` and `UNINSTALL_MANAGED_EXCEPTION_DISPATCHER` to match the initialization order?

janvorli replied:
My mistake, of course it should be there.

janvorli replied:
Hmm, moving it there has broken it somehow, I'll take a look in the morning.

janvorli committed:
Reflect PR feedback

Also fix order of __helperframe.Push() and
MAKE_CURRENT_THREAD_AVAILABLE_EX(__helperframe.GetThread()) that
was causing runtime crash.
